Add flatc build testing against clang

neomantra opened the pull request:
This adds a Dockerfile which tests building flatc against clang.  See discussion at #5119.  It's a copy of the gcc-based Dockerfile with an additional package (`clang`) and the appropriate `CC` environment variables.

vglavnyy commented:
@neomantra 
Is it possible to reuse `travis.yml` as the template (or create a new template with GPRC, sanitizers and `ctest` run)?
https://github.com/google/flatbuffers/blob/63d51afd1196336a7d1f56a988091ef05deb1c62/.travis.yml#L91-L103
With test environment `FLATBUFFERS_BUILD_GRPCTEST=ON`, `DFLATBUFFERS_CODE_SANITIZE=ON` and
https://github.com/google/flatbuffers/blob/63d51afd1196336a7d1f56a988091ef05deb1c62/.travis.yml#L5-L8

neomantra commented:
I've updated these per your suggestions.  I also added a GCC Dockerfile for better coverage.  I think the `ctest` run does all the tests you want -- if not let me know.

neomantra commented:
This failed on Travis because the build took too long:
```
The job exceeded the maximum time limit for jobs, and has been terminated.
```
It looks like this build stage normally take less than 10 minutes.  

With these full C++ / GRPC builds, it is taking much longer and Travis has a limit of 50 minutes.   Each C++ build (there are two, GCC and Clang) takes about 30 minutes.

I will put these into separate Travis stages, rather than being under the "languages" stage...

neomantra commented:
Noting that the Appveyor build failure there is unrelated to this commit.  It was a failure to [find a Rust crate repository](https://ci.appveyor.com/project/gwvo/flatbuffers/builds/21816486/job/4t0y0v3sgq643h3l).    Note the Travis build passed and it built the Dockerfiles successfully.

vglavnyy commented:
Probably, the idea to test grpc with all compilers was not good.
The build and test time is 2 hours. This is too long.

https://travis-ci.org/google/flatbuffers/jobs/483157950
```
$ docker build -f tests/docker/Dockerfile.testing.cpp.clang_3_8_1 .
-- The C compiler identification is Clang 3.8.1
-- The CXX compiler identification is GNU 6.3.0
```
`GNU 6.3.0` is GCC 6.3.0.
Is it right that `CXX compiler identification is GNU 6.3.0` with `ENV CCX /usr/bin/clang++`?

`tests/docker/Dockerfile.testing.cpp.clang_3_8_1` looks correct:
```
ENV UBSAN_OPTIONS "halt_on_error=1"
ENV ASAN_OPTIONS "halt_on_error=1"
ENV CC /usr/bin/clang
ENV CCX /usr/bin/clang++
```

neomantra commented:
Good catch... typo there in the Dockerfile ENV statement : `CCX` vs `CXX`.   I am running a test build again locally and will update once it is straight.

With the same commit, I will make it faster by adding parallel build of 2.  This should be OK for Travis, [according to this](https://docs.travis-ci.com/user/reference/overview/#virtualization-environments). 
 Maybe that will shorten it by half.

With respect to whether this is part of the build or not, is up to the maintainers.   If not accepted, there's still some coverage from my weekly Docker build.

neomantra commented:
I have added parallel builds with `JOBS=2`.  I also restructured the Dockerfiles a built so that the build caching mechanisms can speed things up a bit (mostly building GRPC).  Hopefully this cuts down on the build time.

I also fixed the `CCX -> CXX` typo.

neomantra commented:
I removed the Docker GCC build from the `.travis.yml` (but left the corresponding Dockerfile) because it largely does the same thing as the co-existing Travis builds (the stanza below what I just removed).  So in the end this changeset just adds the clang build and 20-30 minutes of build time.

vglavnyy commented on tests/docker/Dockerfile.testing.cpp.gcc_6_3_0:
```diff
@@ -0,0 +1,28 @@
+ENV UBSAN_OPTIONS "halt_on_error=1"
+ENV ASAN_OPTIONS "halt_on_error=1"
+ADD grpc/build_grpc.sh grpc/build_grpc.sh
+RUN bash grpc/build_grpc.sh
```
Remove (comment) protobuf compilation while GRPC doesn't work (#5099)?

neomantra commented:
I don't think the Appveyor build failure is related to any of my changes.  It looks like the build was hung during compilation and was terminated by Appveyor after an hour.

rw commented:
@neomantra I think you'll want to put the Docker files in the `languages` subdir so that the CI script picks them up: https://github.com/google/flatbuffers/tree/master/tests/docker/languages

neomantra commented:
@rw I took it out of that directory because Travis build steps are limited to one hour and that whole `languages` subdir build is one step.  Adding the ~30 minute C++ build to that step caused the timeout to happen.

Thus I separated it into its own Travis step (in `.travis.yml`) with the comment:
```
# run separately from build-and-run-docker-test-containers.sh because of build timeouts
```

vglavnyy commented:
@aardappel 
Do you have a doubt about this pull-request?
This PR will enable testing Flatbuffers with the latest versions of GCC and Clang.
I checked this PR with d8210d5a83faa345046648e520d82b54ea724e35, all CI passed.

neomantra commented:
To clarify, because of build time and that the build already does a non-Docker GCC, the final squashed commit of this adds the Dockerfile for GCC but does not run it. 

It does run the Clang one, but not via the `languages` subdir but rather as an explicit build step.    

There were a bunch of updates to this PR to get it going but the squashed diff is pretty simple.

aardappel commented:
@vglavnyy No, seems like a good idea to me.
And yes, optimizing CI such that they don't time out so often would be good :)

aardappel commented:
@rw can merge once he's happy :)

rw commented:
Can we get CI to pass with a bump? :-]

neomantra commented:
I rebased this to master and squashed this PR.

The CI build is failing on clang because of https://github.com/google/flatbuffers/commit/9bf9b18f0a705dfd6d50b98056463a55de6a1bf9

aardappel commented:
@neomantra apologies.. just pushed a fix for that.

neomantra commented:
Clean build 🎉

vglavnyy commented:
@rw merge?
After the merge, we can move Clang version to 7.0 or newer (#5258).
